### PR TITLE
cmd/roachtest: add disagg-rebalance roachtest

### LINKED ISSUE #103030: roachtest: Add new roachtest to test fast rebalance
Once #103028 is complete, a roachtest that creates a cluster, loads a fixture, and adds a node or two (and ensures they catch up in replica count without crashing) would be good to have as an end-to-end test for disaggregated ingestions / fast rebalances.

Jira issue: CRDB-27802

### PR DESCRIPTION
**Pull request by itsbilal**
This test adds a roachtest that spins up a cluster with 3 nodes using S3 as the --experimental-shared-storage, and then adds a fourth node after loading a tpcc fixture and with a foreground workload running on it. It confirms the fourth node gets hydrated without transferring all live bytes over the wire.

Epic: none
Fixes: #103030

Release note: None

### CODE REVIEW

**Comment by itsbilal** [comment]
Tested this with #105839 and it reliably passes. Tested it without that change and one of the assertions fails. Won't merge this until that PR is merged.

**Review by RaduBerinde — APPROVED** [review]
:lgtm: Nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-:-N_uiLAQ2WCUUEiWk_O0:b-8gk8uj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @renatolabs and @srosenberg)

**Review by srosenberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-N_uxrpKBSZ7pK-V88eT:-N_uxrpKBSZ7pK-V88eU:b-pb2edg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e861bae1c2648a37026db75841cbe33c541833/pkg/cmd/roachtest/tests/disagg-rebalance.go#L41)):*
> ```Go
> 			c.Put(ctx, t.Cockroach(), "./cockroach")
> 			c.Put(ctx, t.DeprecatedWorkload(), "./workload")
> 			s3dir := "s3://cockroachdb-backup-testing/disagg-rebalance/" + c.Name() + "?AUTH=implicit"
> ```

Please use `testutils.BackupTestingBucket()` [1]. @renatolabs Need to think how to surface this better; we also need to start discouraging (maybe a linter?) test authors from using any hardcoded buckets.

[1] https://github.com/cockroachdb/cockroach/pull/107073

**Review by srosenberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @renatolabs)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-N_v0DXB269-lTEL4YLu:-N_v0DXB269-lTEL4YLv:bnki8oy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e861bae1c2648a37026db75841cbe33c541833/pkg/cmd/roachtest/tests/disagg-rebalance.go#L54)):*
> ```Go
> 				warehouses,
> 			)}
> 			c.Run(ctx, c.Node(1), cmd...)
> ```

Please use `c.NewMonitor(ctx, c.Range(1, 3))` [1], this way we catch silent deaths. The pattern is typically as follows,

```
m := c.NewMonitor(ctx, crdbNodes)
...
m.Go(doSomething())
...
m.Go(doSomethingElse())
...
m.Wait()
```

@renatolabs I wonder if we should make it mandatory? i.e., every time a test starts a cluster, the (roachtest) API automatically adds a monitor.

[1] https://github.com/cockroachdb/cockroach/pull/36224

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-:-Na7UCG820gMnhb7V3c0:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @renatolabs)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-N_uxrpKBSZ7pK-V88eT:-Na7UDHv9tXw7LQASfG9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e861bae1c2648a37026db75841cbe33c541833/pkg/cmd/roachtest/tests/disagg-rebalance.go#L41)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Please use `testutils.BackupTestingBucket()` [1]. @renatolabs Need to think how to surface this better; we also need to start discouraging (maybe a linter?) test authors from using any hardcoded buckets.

[1] https://github.com/cockroachdb/cockroach/pull/107073
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-N_v0DXB269-lTEL4YLu:-Na7UDnE6Zef-oqSTwlE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e861bae1c2648a37026db75841cbe33c541833/pkg/cmd/roachtest/tests/disagg-rebalance.go#L54)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

Please use `c.NewMonitor(ctx, c.Range(1, 3))` [1], this way we catch silent deaths. The pattern is typically as follows,

```
m := c.NewMonitor(ctx, crdbNodes)
...
m.Go(doSomething())
...
m.Go(doSomethingElse())
...
m.Wait()
```

@renatolabs I wonder if we should make it mandatory? i.e., every time a test starts a cluster, the (roachtest) API automatically adds a monitor.

[1] https://github.com/cockroachdb/cockroach/pull/36224
</blockquote></details>

Done.

**Comment by itsbilal** [comment]
Will merge this after #105839 merges.

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7YE2-3CAJZIqKJ1B3:-Na7YE206H-6kaT7rpOm:b-sqdyyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L30)):*
> ```Go
> func registerDisaggRebalance(r registry.Registry) {
> 	disaggRebalanceSpec := r.MakeClusterSpec(4)
> 	r.Add(registry.TestSpec{
> ```

You need to add the aws tag for this to be picked up by the nightly build.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7Yrpv8ciSN0diIVtL:-Na7Yrpv8ciSN0diIVtM:bwpvhib) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L51)):*
> ```Go
> 
> 			t.Status(`workload initialization`)
> 			cmd := []string{fmt.Sprintf(
> ```

Nit: this can be a plain string.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7YZe16qB-r0STPxVC:-Na7YZe16qB-r0STPxVD:b-b990hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L52)):*
> ```Go
> 			t.Status(`workload initialization`)
> 			cmd := []string{fmt.Sprintf(
> 				"./workload fixtures import tpcc --warehouses=%d {pgurl:1}",
> ```

We can replace the uses of `./workload` in this test with `./cockroach workload`. As an added bonus, we'll then no longer need to upload the `workload` binary to all nodes.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7Z6J40TqFxDtU14V0:-Na7Z6J40TqFxDtU14V1:b5ahtqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L61)):*
> ```Go
> 			m.Wait()
> 
> 			ctx, cancel := context.WithCancel(ctx)
> ```

Do we need a child context? It is only canceled when the function returns, so the waits for `<-ctx.Done()` below would only trigger if the parent context canceled. If there _is_ a reason for it that I missed, a comment should make it clearer.

If we don't need it, it's probably better to remove this, since it can get kind of confusing with all the shadowing (`ctx` passed to `Run`, this local `ctx`, then `ctx` passed to `monitor.Go`.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_49cDmV6rDJAkeo6:-Na7_49cDmV6rDJAkeo7:bayaexd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L64)):*
> ```Go
> 			defer cancel()
> 			cmdDone := make(chan error)
> 			m2 := c.NewMonitor(ctx, c.Range(1, 3))
> ```

Any reason why we need a new monitor? I thought they could be reused.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_bfH1tm3VDnP735g:-Na7_bfI8YIzo2QpCPLR:bc13wj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L74)):*
> ```Go
> 
> 				err := c.RunE(ctx, c.Node(1), cmd)
> 				cmdDone <- err
> ```

If I'm reading this test right, if one of the assertions below fail (e.g., one of the `db.QueryRow`s), the test would fail and this goroutine will stay blocked in this channel send. I believe we can fix this by changing the wait loop at the end of this test, see my comment there.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 97 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_9FA7qaKbBF2mMl1:-Na7_9FA7qaKbBF2mMl2:b-68d3cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L97)):*
> ```Go
> 			if err := waitForRebalance(ctx, t.L(), db, 10 /* maxStdDev */, 20 /* stableSeconds */); err != nil {
> 				t.Fatal(err)
> 				return
> ```

Nit: unreachable.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 107 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_BD-4Jyv-BpC8XyX:-Na7_BD-4Jyv-BpC8XyY:b-68d3cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L107)):*
> ```Go
> 			).Scan(&count); err != nil {
> 				t.Fatal(err)
> 				return
> ```

Nit: unreachable.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_rNI7SkI_Yn5b89R:-Na7_rNI7SkI_Yn5b89S:b-vtt0vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L143)):*
> ```Go
> 				return
> 			}
> 			m2.Wait()
> ```

Couldn't this `select` block be replaced by just a Wait on the monitor? Then we can even do away with `cmdDone`. If the context is canceled (most likely reason would be a test timeout), `Wait` should return since the `RunE` should observe that cancelation as well and return.

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-N_uxrpKBSZ7pK-V88eT:-Na7awFAAEd-jYR_PxTo:bn27w4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e861bae1c2648a37026db75841cbe33c541833/pkg/cmd/roachtest/tests/disagg-rebalance.go#L41)):*
> Need to think how to surface this better; we also need to start discouraging (maybe a linter?) test authors from using any hardcoded buckets.

Agree, but not sure what the best approach would be. A general purpose linter is tricky, and there would probably be false positives.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-N_v0DXB269-lTEL4YLu:-Na7bMXiDZ9Mf5RgbO3a:bw1pvdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e861bae1c2648a37026db75841cbe33c541833/pkg/cmd/roachtest/tests/disagg-rebalance.go#L54)):*
> I wonder if we should make it mandatory? i.e., every time a test starts a cluster, the (roachtest) API automatically adds a monitor.

Not a bad idea; we would then also need to lift the `ExpectDeath` API to the (already crowded) `cluster` interface. IMO, most tests would fail if a node suddenly crashed, but the error message would be less direct than `unexpected node death`.

What we _could_ add a linter for would be using `go` in roachtests.

**Review by itsbilal — COMMENTED** [review]
TFTR!

> Dismissed @srosenberg from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-:-NaCf15d8_d6f5O2errp:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7YE2-3CAJZIqKJ1B3:-NaCd-lo2hiFJ9zZ5ETv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L30)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

You need to add the aws tag for this to be picked up by the nightly build.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7Yrpv8ciSN0diIVtL:-NaCdKpN8C8QuOLDJLqb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L51)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: this can be a plain string.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7YZe16qB-r0STPxVC:-NaCdMsH0Qyb4Uthwym6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L52)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

We can replace the uses of `./workload` in this test with `./cockroach workload`. As an added bonus, we'll then no longer need to upload the `workload` binary to all nodes.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7Z6J40TqFxDtU14V0:-NaCdTXq6MUpj0hhTpWM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L61)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Do we need a child context? It is only canceled when the function returns, so the waits for `<-ctx.Done()` below would only trigger if the parent context canceled. If there _is_ a reason for it that I missed, a comment should make it clearer.

If we don't need it, it's probably better to remove this, since it can get kind of confusing with all the shadowing (`ctx` passed to `Run`, this local `ctx`, then `ctx` passed to `monitor.Go`.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_49cDmV6rDJAkeo6:-NaCdTnm4Wonc9Vr-xIm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L64)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Any reason why we need a new monitor? I thought they could be reused.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_bfH1tm3VDnP735g:-NaCeydM0MSOci_mweIY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L74)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

If I'm reading this test right, if one of the assertions below fail (e.g., one of the `db.QueryRow`s), the test would fail and this goroutine will stay blocked in this channel send. I believe we can fix this by changing the wait loop at the end of this test, see my comment there.
</blockquote></details>

Done.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_rNI7SkI_Yn5b89R:-NaCeF-c4RWDDD5EZFeJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L143)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Couldn't this `select` block be replaced by just a Wait on the monitor? Then we can even do away with `cmdDone`. If the context is canceled (most likely reason would be a test timeout), `Wait` should return since the `RunE` should observe that cancelation as well and return.
</blockquote></details>

Done.

**Review by itsbilal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7YZe16qB-r0STPxVC:-NaCskGpBC-EZ7sIrLYO:b-jy0dmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L52)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Ended up reverting this, for some reason it was erroring out with `cockroach workload` but not with `workload` ?
</blockquote></details>

nevermind, the error was due to monitor reuse. Brought back `cockroach workload`.
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_49cDmV6rDJAkeo6:-NaCs_Kq7wTV7Nd8X5gq:b-lb6wns) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L64)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Had to revert this; two monitors are necessary because `Wait()` cancels its context so calling Go() after a wait is not allowed.

**Review by renatolabs — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-:-NaCxj0r0pGh-qCry0wp:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7Yrpv8ciSN0diIVtL:-NaCxOl52rT7XIpXrH8O:bfnbjyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L51)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I still see it as `[]string`?
___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7_49cDmV6rDJAkeo6:-NaCxTqV95ijJNp68uRd:by3nrnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L64)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Had to revert this; two monitors are necessary because `Wait()` cancels its context so calling Go() after a wait is not allowed.
</blockquote></details>

Ah yes, makes sense. Sorry for the misleading comment!
___
*[`pkg/cmd/roachtest/tests/disagg_rebalance.go` line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-NaCxJZeCh9ZTi97IJez:-NaCxJZf5BSp0TxHVCD0:bg5c42v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b18f99f4c6aad93158c98e5250f20e47a61aef1/pkg/cmd/roachtest/tests/disagg_rebalance.go#L42)):*
> ```Go
> 			}
> 			c.Put(ctx, t.Cockroach(), "./cockroach")
> 			c.Put(ctx, t.DeprecatedWorkload(), "./workload")
> ```

Nit: no longer needed.

**Review by itsbilal — COMMENTED** [review]
TFTR!

Gonna wait for the dependent changes to merge before I merge this, which could take ~2 weeks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-:-NaD0IDp10v4mcrvbuFz:b-x7578b)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/tests/disagg-rebalance.go` line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/107394#-Na7Yrpv8ciSN0diIVtL:-NaD0Oax6BFN6hHBvtMm:b-jxqmnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0e242c21ca33a5bd88a8fb5d8509e4550f9327e/pkg/cmd/roachtest/tests/disagg-rebalance.go#L51)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I still see it as `[]string`?
</blockquote></details>

Ah I thought  you were referring to the t.Status string above which was previously a raw string literal. Fixed the []string now.

**Comment by itsbilal** [comment]
bors r=renatolabs

**Pull request merged**
